docs/contributing-guide/testing/01-unit-tests.md

```diff
@@ -26,3 +26,208 @@ For unit testing we use [`vitest`](https://vitest.dev/guide/), to run the unit t
 ```shell
 npm test
 ```
+
+## Principles
+
+1. **No mocking**
+   - Use the real Workers test runtime (D1, KV, R2, timers). Storage is isolated per test.
+2. **Do not test Zod/validators**
+   - Validation libraries are well-tested. Avoid re-testing them.
+3. **Types from schemas**
+   - Derive types with `z.infer<typeof Schema>` (or Valibot/TypeBox equivalents). Use `Pick`/`Omit`/`Partial` when necessary.
+4. **Focus on post-validation logic**
+   - Validation problems are already caught by Hurl/API tests. Unit tests assert business logic and side effects **after** validation.
+
+## Writing classic unit tests
+
+A "pure" function example (no side effects):
+
+```ts
+// src/math.ts
+export function clamp(n: number, lo: number, hi: number) {
+	if (Number.isNaN(n)) {
+		throw new TypeError('n must be a number');
+	}
+
+	return Math.min(hi, Math.max(lo, n));
+}
+```
+
+```ts
+// tests/math.spec.ts
+import { describe, expect, it } from 'vitest';
+import { clamp } from '../src/math';
+
+describe('clamp()', () => {
+	it('handles integers, floats, +0 and -0', () => {
+		expect(clamp(3, 0, 5)).toBe(3);
+		expect(clamp(3.3, 0, 5)).toBe(3.3);
+		expect(Object.is(clamp(+0, -1, 1), +0)).toBe(true);
+		expect(Object.is(clamp(-0, -1, 1), -0)).toBe(true);
+	});
+
+	it('clips at bounds and negatives', () => {
+		expect(clamp(-10, -5, 5)).toBe(-5);
+		expect(clamp(99, -5, 5)).toBe(5);
+	});
+
+	it('handles big and special values', () => {
+		expect(clamp(Number.MAX_SAFE_INTEGER + 1, 0, Number.MAX_SAFE_INTEGER))
+			.toBe(Number.MAX_SAFE_INTEGER);
+		expect(clamp(Infinity, 0, 1)).toBe(1);
+		expect(() => clamp(NaN as any, 0, 1)).toThrow(/number/);
+	});
+});
+```
+
+## Tests that use real Workers runtime
+
+### Test APIs & storage isolation
+
+Vitest exposes bindings and helpers from `cloudflare:test`.
+
+```ts
+// test/setup.ts
+import { env } from 'cloudflare:test';
+
+// Example: ensure D1 schema exists before tests
+export default async function() {
+	await env.Database.prepare('CREATE TABLE IF NOT EXISTS audit (sid TEXT, userId TEXT)').run();
+}
+```
```

Comment:
Why is it done this way, any caveats that need to be stated for people not familiar with the codebase?